### PR TITLE
[jdk] Don't require `javac`/`jar` in $PATH

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Windows machines do not include Java in their path by default, so it is
a better experience to not require it for the build.

Changes to make this happen:
- `JavaCPath` and `JarPath` are configured in
`Configuration.OperatingSystem.props`
- Usage of `javac` and `jar` across the repo use appropriate variables
now
- `generate-os-info` needs to set `JavaCPath` and `JarPath`
- Created a new `<Ant>` MSBuild task
- A couple places, Windows needs `JAVA_HOME` to be set, so we are doing
this via `Environment.SetEnvironmentVariable` in a couple MSBuild tasks
- `Configuration.OperatingSystem.props` is conditional, so it is
possible to build xa-prep-tasks without it
- `Which` needs to accept files without extensions last for Windows, `PATHEXT` should be empty on Unix

### CODE REVIEW

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JavaPaths.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+			try {
+				AndroidSdk.Refresh (AndroidSdkPath, AndroidNdkPath, JavaSdkPath);
+
+				var javaSdkPath = AndroidSdk.JavaSdkPath;
+				if (string.IsNullOrEmpty (javaSdkPath)) {
+					Log.LogError ("JavaSdkPath is blank");
+					return false;
+				}
+
+				Log.LogMessage (MessageImportance.Low, $"  {nameof (AndroidSdk.JavaSdkPath)}: {javaSdkPath}");
```
Comment (jonpryor) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JavaPaths.cs]: Current convention (e.g. from `<BuildApk/>`) is that output values should be logged with `[Output]`:

```csharp
Log.LogMessage (MessageImportance.Low, $"  {nameof (JavaC)}: {JavaC}");
Log.LogMessage (MessageImportance.Low, $"  {nameof (Jar)}: {Jar}");
```

I realize that `javaSdkPath` isn't an output variable, but it can also be inferred from the `JavaC` path.

**Comment by jonpryor** [comment]
My biggest problem with this patch is naming-based: the `<JavaPaths/>` task, as a *side-effect*, sets the `$JAVA_HOME` environment variable, but this isn't obvious from the name. It thus results in "oddities" such as in the `_AcceptAndroidSdkLicenses` target where `<JavaPaths/>` is called but no outputs are recorded, nor is it obvious why `_AcceptAndroidSdkLicenses` needs to call `<JavaPaths/>`. (Does `<AcceptAndroidSdkLicenses/>` require that `$JAVA_HOME` be set? If it does, why doesn't it set it itself? etc.)

I'm not convinced that renaming `<JavaPaths/>` to e.g. `<SetJavaPaths/>` helps either, because then the `<Output/>` variables don't really make sense; why do they then exist?

In short, this feels "too magical" to me. :-(

I would instead suggest:

1. Renaming `<JavaPaths/>` to `<GetJava̦Paths/>`.
2. `<GetJavaPaths/>` *does not* call `Environment.SetEnvironmentVariable()`.
3. Add a new `<SetEnvironmentVariable/>` task, which *does* call `Environment.SetEnvironmentVariable()`.
4. Existing tasks which require an environment variable should themselves set it.

**Comment by jonpryor** [comment]
*Alternate* suggestion:

Instead of e.g. `<Exec Command="&quot;$(_JavaC) ..." />`, perhaps `Xamarin.Android.Tools.BootstrapTasks`/etc. should provide `<JavaC/>` and `<Jar/>` tasks, turning this:

```xml
<Exec
    Command="&quot;$(_Jar)&quot; cf &quot;$(_MonoAndroidJar)&quot; -C &quot;$(IntermediateOutputPath)jcw\bin&quot; ."
/>
```

into:

```xml
<Jar
    ToolPath="$(JavaSdkDirectory)"
    ToolExe="$(JarToolExe)"
    Options="cf &quot;$(_MonoAndroidJar)&quot; -C &quot;$(IntermediateOutputPath)jcw\bin&quot; ."
/>
```

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs [hunk 2]
```diff
@@ -40,6 +44,12 @@ public override bool Execute ()
 
 				Log.LogMessage (MessageImportance.Low, $"  {nameof (AndroidSdk.JavaSdkPath)}: {javaSdkPath}");
 
+				if (File.Exists (ConfigurationOperatingSystemProps.ItemSpec)) {
```
Comment (jonathanpeppers) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs]: This is one spot I did something strange: 
1. `Configuration.OperatingSystem.props` has to _exist_ for xa-prep-tasks to build
1. A custom MSBuild task in xa-prep-tasks needs to replace `@JAVA_HOME@` in `Configuration.OperatingSystem.props`, I put this in `JdkInfo` because it was doing the same thing for Java.Interop

Maybe there is a better way?

File: build-tools/scripts/Windows-Configuration.OperatingSystem.props [hunk 3]
```diff
@@ -9,5 +9,8 @@
     <HostTriplet64 Condition=" '$(HostTriplet64)' == '' ">x86_64-win32</HostTriplet64>
     <HostCpuCount Condition=" '$(HostCpuCount)' == '' ">$([System.Environment]::ProcessorCount)</HostCpuCount>
     <HostBits Condition=" '$(HostBits)' == '' ">64</HostBits>
+    <JavaSdkDirectory>@JAVA_HOME@</JavaSdkDirectory>
```
Comment (jonpryor) [inline build-tools/scripts/Windows-Configuration.OperatingSystem.props]: If we're going to start replacing `@...@`, please rename the file to have a `.in` extension, to make it clearer that this is a template file.

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs (hunk 2 above)
Comment (jonpryor) [reply build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs]: We already have a task which handles `@...@` replacement: [`<ReplaceFileContents/>`](https://github.com/xamarin/xamarin-android/tree/master/build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ReplaceFileContents.cs). I would suggest reusing that:

```xml
<ReplaceFileContents
    SourceFile="Properties\AssemblyInfo.cs.in"
    DestinationFile="$(IntermediateOutputPath)AssemblyInfo.cs"
    Replacements="@JAVA_HOME@=$(_TheJavaSdkPath)"
/>
```

This simply raises another question: where does `$(_TheJavaSdkPath)` come from? You should make `JdkInfo.JavaSdkPath` an `[Output]` property, which would allow you to set `$(_TheJavaSdkPath)` and use it in a later `<ReplaceFileContents/>` call.

Comment (jonpryor) [reply build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs]: > 1. `Configuration.OperatingSystem.props` has to exist for xa-prep-tasks to build

What we could do is add another conditional import ("Indirections, all the way down!") to `Configuration.OperatingSystem.props`:

```xml
<Import
    Project="$(MSBuildThisFileDirectory)Configuration.Jdk.Props"
    Condition="Exists('$(MSBuildThisFileDirectory)Configuration.Jdk.Props')"
/>
```

You could then place the `$(JavaSdkDirectory)`, `$(JavaCPath)`, and `$(JarPath)` values within `build-tools\scripts\windows-Configuration.Jdk.Props.in`.

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs [hunk 4]
```diff
@@ -67,6 +77,10 @@ public override bool Execute ()
       </ItemGroup>
     </When>
   </Choose>
+  <PropertyGroup>
```
Comment (jonpryor) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/JdkInfo.cs]: ...with the above `<ReplaceFileContents/>` approach in mind, you could *remove* this `File.WriteAlltext()` block here, and replace with an approach of additional `[Output]` properties within `JdkInfo` + a `<ReplaceFileContents/>` call.

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/SetEnvironmentVar.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+
+namespace Xamarin.Android.BuildTools.PrepTasks
+{
+	public class SetEnvironmentVar : Task
```
Comment (jonpryor) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/SetEnvironmentVar.cs]: We have code completion and (occasional) IDEs. We can use `SetEnvironmentVariable`. :-)

File: src/proguard/proguard.targets [hunk 6]
```diff
@@ -24,6 +30,11 @@
   </Target>
   <Target Name="_CleanProGuard"
       AfterTargets="Clean">
+    <SetEnvironmentVar
```
Comment (jonpryor) [inline src/proguard/proguard.targets]: It *looks* like that the primary reason for `<SetEnvironmentVar/>` is for `ant`. Would it be better to have an `<Ant/>` task which calls `ant` and also calls `Environment.SetEnvironmentVariable()`?

```xml
<Ant
    Command="-f buildscripts\build.xml clean"
    ToolPath="$(AntToolPath)"
    ToolExe="$(AntToolExe)"
    JavaSdkPath="$(JavaSdkDirectory)"    
/>
```

**Comment by jonpryor** [comment]
build

File: build-tools/scripts/PrepareWindows.targets [hunk 7]
```diff
@@ -5,28 +5,27 @@
...
-        DestinationFiles="@(_ConfigurationFile->'%(Destination)')"
-        SkipUnchangedFiles="True"
-    />
     <MSBuild Projects="$(MSBuildThisFileDirectory)..\xa-prep-tasks\xa-prep-tasks.csproj" />
     <JdkInfo
         AndroidSdkPath="$(AndroidSdkPath)"
         AndroidNdkPath="$(AndroidNdkPath)"
         JavaSdkPath="$(JavaSdkDirectory)"
-        Output="$(_TopDir)\external\Java.Interop\bin\BuildDebug\JdkInfo.props"
+        Output ="$(_TopDir)\external\Java.Interop\bin\BuildDebug\JdkInfo.props">
```
Comment (jonpryor) [inline build-tools/scripts/PrepareWindows.targets]: The space before the `=` is weird.

File: build-tools/scripts/PrepareWindows.targets [hunk 8]
```diff
@@ -5,28 +5,27 @@
...
+        <Output TaskParameter="JavaSdkDirectory" PropertyName="_JavaSdkDirectory" />
```
Comment (jonpryor) [inline build-tools/scripts/PrepareWindows.targets]: This is indented too much; it should be 2 spaces from parent element column, not 4. (*Attributes* are 4 spaces.)

File: build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/Which.cs [hunk 9]
```diff
@@ -30,10 +30,13 @@ static Which ()
 		{
 			var pathExt     = Environment.GetEnvironmentVariable ("PATHEXT");
 			var pathExts    = pathExt?.Split (new char [] { Path.PathSeparator }, StringSplitOptions.RemoveEmptyEntries);
-			FileExtensions  = new string [(pathExts?.Length ?? 0) + 1];
-			FileExtensions [0] = null;
+			bool isUnix     = Environment.OSVersion.Platform != PlatformID.Win32NT;
+
+			//NOTE: Unix should accept blank extensions first and Windows last
+			FileExtensions = new string [(pathExts?.Length ?? 0) + 1];
+			FileExtensions [isUnix ? 0 : FileExtensions.Length - 1] = null;
```
Comment (jonpryor) [inline build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/Which.cs]: The `isUnix` check isn't necessary. On Unix, `$PATHEXT` will be empty, so `FileExtensions.Length` is only 1.

Instead, just make the null entry last:

```csharp
FileExtensions  = new string [(pathExts?.Length ?? 0) + 1];
if (pathExts != null) {
    Array.Copy (pathExts, 0, FileExtensions, 0, pathExts.Length);
}
FileExtensions [FileExtensions.Length-1] = null;
```

**Comment by jonathanpeppers** [comment]
The `macOS+xbuild` is giving:
```
error: cannot fork() for rev-list: Resource temporarily unavailable
```

Earlier today it happened and I waited a while and tried again and it worked. I think it might mean there are [too many processes running](https://superuser.com/questions/441889/error-cannot-fork-for-status-resource-temporarily-unavailable-git), we could try a reboot?
